sk-csharp-chatgpt-plugin/semantic-functions-generator/SemanticSkillGenerator.cs

```diff
@@ -0,0 +1,177 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using AIPlugins.AzureFunctions.Generator.Extensions;
+using AIPlugins.AzureFunctions.Generator.Models;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Newtonsoft.Json;
+
+namespace AIPlugins.AzureFunctions.Generator;
+
+[Generator]
+public class SemanticFunctionGenerator : ISourceGenerator
+{
+    private const string DefaultFunctionNamespace = "AIPlugins";
+    private const string FunctionConfigFilename = "config.json";
+    private const string FunctionPromptFilename = "skprompt.txt";
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+        var rootNamespace = context.GetRootNamespace();
+
+        if (String.IsNullOrEmpty(rootNamespace))
+        {
+            rootNamespace = DefaultFunctionNamespace;
+        }
+
+        // Get the additional files that represent the functions and functions
+        var functionFiles = context.AdditionalFiles.Where(f =>
+            f.Path.Contains(FunctionConfigFilename) ||
+            f.Path.Contains(FunctionPromptFilename));
+
+        // Group first by function name, then by parent folder
+        var fnFileGroup = functionFiles.GroupBy(f => Path.GetDirectoryName(f.Path));
+
+        // Group the files by parent folder name
+        var folderGroups = fnFileGroup.GroupBy(f => Path.GetFileName(Path.GetDirectoryName(f.Key)));
+
+        // Generate a class for each folder
+        foreach (var folderGroup in folderGroups)
+        {
+            string? folderName = folderGroup.Key;
+            if (string.IsNullOrWhiteSpace(folderName))
+            {
+                continue;
+            }
+
+            string classSource = GenerateClassSource(rootNamespace!, folderName, folderGroup);
+            context.AddSource(folderName, SourceText.From(classSource, Encoding.UTF8));
+        }
+    }
+
+    // Generate the source code for a folder of semantic functions
+    private static string GenerateClassSource(string rootNamespace, string folderName, IGrouping<string, IGrouping<string, AdditionalText>> folderGroup)
+    {
+        // Use a StringBuilder to build the class source
+        StringBuilder functionsCode = new();
+
+        foreach (var functionGroup in folderGroup)
+        {
+            // Get the "skprompt.txt" and "config.json" files for this function
+            AdditionalText? configFile = functionGroup.FirstOrDefault(f => Path.GetFileName(f.Path).Equals(FunctionConfigFilename, StringComparison.InvariantCultureIgnoreCase));
+            AdditionalText? promptFile = functionGroup.FirstOrDefault(f => Path.GetFileName(f.Path).Equals(FunctionPromptFilename, StringComparison.InvariantCultureIgnoreCase));
+            if (promptFile != default && configFile != default)
+            {
+                functionsCode.AppendLine(GenerateFunctionSource(promptFile, configFile) ?? string.Empty);
```

Comment:
If we get a `null` response we still proceed as if everything is working. We should probably throw an exception and fail in a way the user knows what is gone wrong and knows what to fix.